---
title: "Rsp_SD_totRNA"
author: "Logan Edvalson"
date: "2025-10-14"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
individualization_GeneSet <- read.table("Individualization_GeneSet.txt")
colnames(individualization_GeneSet) <- c("SYMBOL", "FLYBASE")
individualization_GeneSet <- bitr(individualization_GeneSet$FLYBASE, fromType = "FLYBASE", toType = "SYMBOL", OrgDb = org.Dm.eg.db)
interestingGenes_totRNA <- individualization_GeneSet$SYMBOL
```

```{r}
#PANGEA Background
R16_1 <- read.table("totRNA_Counts/R16-Gla-1.simplified.counts_fam")
R16_2 <- read.table("totRNA_Counts/R16-Gla-2.simplified.counts_fam")
R16_3 <- read.table("totRNA_Counts/R16-Gla-3.simplified.counts_fam")

testis_Background <- data.frame(Gene = unique(c(R16_1[R16_1$V2 > 0 & grepl("^FBgn",R16_1$V1) == TRUE,1],R16_2[R16_2$V2 > 0 & grepl("^FBgn",R16_2$V1) == TRUE,1] ,R16_2[R16_2$V2 > 0 & grepl("^FBgn",R16_2$V1) == TRUE,1])))

write_tsv(testis_Background, "TestisBackground.tsv")
```

```{r}
###############SDMad/Gla vs R16/Gla#############################
countDir <- c("totRNA_Counts/HTSeq_Genes/")
groups <- c(rep("R16", 3), rep("SDMad", 3))
batch <- c(rep("batch3", 6))
files <- paste0(countDir, c("R16-Gla-1_stranded.counts","R16-Gla-2_stranded.counts","R16-Gla-3_stranded.counts","SDMad-Gla-1_stranded.counts","SDMad-Gla-2_stranded.counts","SDMad-Gla-3_stranded.counts"))
countMatrix <- load_counts(files, groups, filter_pattern = "\\)n|dodeca|undeca|rDNA|^_")
countMatrix <- countMatrix[grepl("^FBgn", rownames(countMatrix)) == TRUE, ]

DESeq2_DE(countMatrix, EXP = "SDMad", CTRL = "R16", 
          COLOR = "seagreen", DOWNCOLOR = "grey", TYPE = "totRNA", batch = batch, BACKGROUND = "Gla", plotWidth = 9, plotHeight = 6)

write_clip(na.omit(SDMad_R16_totRNA_Gla$Name[SDMad_R16_totRNA_Gla$p.Adj < 0.01 & grepl("^FBgn", SDMad_R16_totRNA_Gla$Name) == TRUE]))

SDMadvR16_Gla_DotPlotData <- read_csv("PANGEA_Results/SDMadvR16_Gla_PANGEA.csv")
WGCNA_PangeaDotPlot(SDMadvR16_Gla_DotPlotData, dotplotprefix = "SDMadvR16_Gla", numSets = 15)

countDir <- c("totRNA_Counts/PropCount_Repeats/")
groups <- c(rep("R16", 3), rep("SDMad", 3))
batch <- c(rep("batch3", 6))
files <- paste0(countDir, c("R16-Gla-1.out_fam","R16-Gla-2.out_fam","R16-Gla-3.out_fam","SDMad-Gla-1.out_fam","SDMad-Gla-2.out_fam","SDMad-Gla-3.out_fam"))
countMatrix <- load_counts(files, groups, filter_pattern = "\\)n|dodeca|undeca|rDNA|^_")
countMatrix <- countMatrix[grepl("^FBgn", rownames(countMatrix)) == FALSE, ]

DESeq2_DE(countMatrix, EXP = "SDMad", CTRL = "R16", 
          COLOR = "seagreen", DOWNCOLOR = "grey", TYPE = "totRNA_Repeats", BACKGROUND = "Gla", batch = batch, plotWidth = 9, plotHeight = 6)

###############SDMad/Iso1 vs R16/Iso1#############################
countDir <- c("totRNA_Counts/HTSeq_Genes/")
groups <- c(rep("R16", 3), rep("SDMad", 2))
files <- paste0(countDir, c("R16cnbw_S21_L008_stranded.counts","R16cnbw1_S73_L004_stranded.counts","R16cnbw2_S77_L004_stranded.counts","SDcnbw_S22_L008_stranded.counts","SDcnbw_2_S76_L004_stranded.counts"))
batch <- c("batch1", "batch2", "batch2", "batch1", "batch2")
countMatrix <- load_counts(files, groups, filter_pattern = "\\)n|dodeca|undeca|rDNA|^_", miRNA_list = miRNA_List)
countMatrix <- countMatrix[grepl("^FBgn", rownames(countMatrix)) == TRUE, ]

DESeq2_DE(countMatrix, EXP = "SDMad", CTRL = "R16", 
          COLOR = "seagreen", DOWNCOLOR = "grey", TYPE = "totRNA", BACKGROUND = "Iso1", batch = batch, plotWidth = 9, plotHeight = 6)


countDir <- c("totRNA_Counts/PropCount_Repeats/")
groups <- c(rep("R16", 3), rep("SDMad", 2))
files <- paste0(countDir, c("R16cnbw_S21_L008.out_fam","R16cnbw1_S73_L004.out_fam","R16cnbw1_S73_L004.out_fam","SDcnbw_S22_L008.out_fam","SDcnbw_2_S76_L004.out_fam"))
batch <- c("batch1", "batch2", "batch2", "batch1", "batch2")
countMatrix <- load_counts(files, groups, filter_pattern = "\\)n|dodeca|undeca|rDNA|^_", miRNA_list = miRNA_List)
countMatrix <- countMatrix[grepl("^FBgn", rownames(countMatrix)) == FALSE, ]

DESeq2_DE(countMatrix, EXP = "SDMad", CTRL = "R16", 
          COLOR = "seagreen", DOWNCOLOR = "grey", TYPE = "totRNA_Repeats", BACKGROUND = "Iso1", batch = batch, plotWidth = 9, plotHeight = 6)

```

```{r}
###############SD5/Gla vs R16/Gla###############################
countDir <- c("totRNA_Counts/HTSeq_Genes/")
groups <- c(rep("R16", 3), rep("SD5", 3))
files <- paste0(countDir, c("R16-Gla-1_stranded.counts","R16-Gla-2_stranded.counts","R16-Gla-3_stranded.counts","SD5-Gla-1_stranded.counts","SD5-Gla-2_stranded.counts","SD5-Gla-3_stranded.counts"))
batch <- c("batch3","batch3","batch3","batch3","batch3","batch3")
countMatrix <- load_counts(files, groups, filter_pattern = "\\)n|dodeca|undeca|rDNA|^_", miRNA_list = miRNA_List)
countMatrix <- countMatrix[grepl("^FBgn", rownames(countMatrix)) == TRUE, ]

DESeq2_DE(countMatrix, EXP = "SD5", CTRL = "R16", 
          COLOR = "dodgerblue3", DOWNCOLOR = "grey", TYPE = "totRNA", BACKGROUND = "Gla", batch = batch, plotWidth = 9, plotHeight = 6)

write_clip(na.omit(SD5_R16_totRNA_Gla$Name[SD5_R16_totRNA_Gla$p.Adj < 0.01 & grepl("^FBgn", SD5_R16_totRNA_Gla$Name) == TRUE]))

SD5vR16_Gla_DotPlotData <- read_csv("PANGEA_Results/SD5vR16_Gla_PANGEA.csv")
WGCNA_PangeaDotPlot(SD5vR16_Gla_DotPlotData, dotplotprefix = "SD5vR16_Gla", numSets = 15)

countDir <- c("totRNA_Counts/PropCount_Repeats/")
groups <- c(rep("R16", 3), rep("SD5", 3))
files <- paste0(countDir, c("R16-Gla-1.out_fam","R16-Gla-2.out_fam","R16-Gla-3.out_fam","SD5-Gla-1.out_fam","SD5-Gla-2.out_fam","SD5-Gla-3.out_fam"))
batch <- c("batch3","batch3","batch3","batch3","batch3","batch3")
countMatrix <- load_counts(files, groups, filter_pattern = "\\)n|dodeca|undeca|rDNA|^_", miRNA_list = miRNA_List)


DESeq2_DE(countMatrix, EXP = "SD5", CTRL = "R16", 
          COLOR = "dodgerblue3", DOWNCOLOR = "grey", TYPE = "totRNA_Repeats", BACKGROUND = "Gla", batch = batch, plotWidth = 9, plotHeight = 6)


###############SD5/Iso1 vs R16/Iso1#############################
countDir <- c("totRNA_Counts/HTSeq_Genes/")
groups <- c(rep("R16", 3), rep("SD5", 3))
files <- paste0(countDir, c("R16cnbw_S21_L008_stranded.counts","R16cnbw1_S73_L004_stranded.counts", "R16cnbw2_S77_L004_stranded.counts", "SD5-Iso1-1_stranded.counts", "SD5-Iso1-2_stranded.counts", "SD5-Iso1-3_stranded.counts"))
countMatrix <- load_counts(files, groups, filter_pattern = "\\)n|dodeca|undeca|rDNA^_", miRNA_list = miRNA_List)
countMatrix <- countMatrix[grepl("^FBgn", rownames(countMatrix)) == TRUE, ]
batch <- c("batch1", "batch2", "batch2", "batch4", "batch4", "batch4")
DESeq2_DE(countMatrix, EXP = "SD5", CTRL = "R16", 
          COLOR = "dodgerblue3", DOWNCOLOR = "grey", TYPE = "totRNA", BACKGROUND = "Iso1", batch = NULL, plotWidth = 9, plotHeight = 6)

countDir <- c("totRNA_Counts/PropCount_Repeats/")
groups <- c(rep("R16", 3), rep("SD5", 3))
files <- paste0(countDir, c("R16cnbw_S21_L008.out_fam","R16cnbw1_S73_L004.out_fam", "R16cnbw2_S77_L004.out_fam", "SD5-Iso1-1.out_fam", "SD5-Iso1-2.out_fam", "SD5-Iso1-3.out_fam"))
countMatrix <- load_counts(files, groups, filter_pattern = "\\)n|dodeca|undeca|rDNA^_", miRNA_list = miRNA_List)

DESeq2_DE(countMatrix, EXP = "SD5", CTRL = "R16", 
          COLOR = "dodgerblue3", DOWNCOLOR = "grey", TYPE = "totRNA_Repeats", BACKGROUND = "Iso1", batch = NULL, plotWidth = 9, plotHeight = 6)
```


## Total RNA Weighted Gene Correlation Network Analysis

### Normalize Counts

```{r}
countDir <- c("totRNA_Counts/HTSeq_Genes/")

#################Read in Count Data################################
groups <- c(rep("R16", 3), rep("SD5", 3), rep("SDMad", 3))
files <- paste0(countDir, list.files(countDir)[grepl("Gla-", list.files(countDir))])
countMatrix <- load_counts(files, groups, filter_pattern = "\\)n|dodeca|undeca|rDNA", miRNA_list = NULL)

if (is.null(colnames(countMatrix)) || anyDuplicated(colnames(countMatrix)) > 0) {
    colnames(countMatrix) <- paste0(groups, "_", seq_along(groups))
  }
  # Build colData with your conditions
  colData <- data.frame(condition = factor(groups))
  rownames(colData) <- colnames(countMatrix)
  # Set CTRL as the reference level
  colData$condition <- relevel(colData$condition, ref = "R16")
  # Build DESeq2 object
  dds <- DESeqDataSetFromMatrix(
    countData = countMatrix,
    colData = colData,
    design = ~ condition
  )
#########Run DESeq2################
dds <- dds[rowSums(counts(dds)) >= 10, ]
dds <- DESeq(dds)

vsd <- varianceStabilizingTransformation(dds)

wpn_vsd <- getVarianceStabilizedData(dds)
rv_wpn <- rowVars(wpn_vsd)
summary(rv_wpn)

q50_wpn <- quantile(rowVars(wpn_vsd), 0.5) 
expr_normalized <- wpn_vsd[rv_wpn > q50_wpn,] 

expr_normalized_df <- data.frame(expr_normalized) %>%
  mutate(
    Gene_id = row.names(expr_normalized)
  ) %>%
  pivot_longer(-Gene_id)

############Normalized Expression Violin Plot at selected quantile###########
expr_normalized_df %>% ggplot(., aes(x = name, y = value)) +
  geom_violin() +
  geom_point() +
  theme_bw() +
  theme(
    axis.text.x = element_text( angle = 90)
  ) +
  ylim(0, NA) +
  labs(
    title = "Normalized and 50 quantile Expression",
    x = "treatment",
    y = "normalized expression"
  )
```

### Select Soft Power

```{r}
input_mat <- t(expr_normalized)

powers <- c(c(1:10), seq(from = 12, to = 50, by = 2))

sft <- pickSoftThreshold(input_mat, powerVector = powers, verbose = 5)

###############Plots to help choose best power (maximize connectivity and r^2 Value###############)
par(mfrow = c(1,2));
  cex1=0.9;
  plot(sft$fitIndices[,1], 
       -sign(sft$fitIndices[,3]) * sft$fitIndices[,2], 
       xlab = "Soft Threshold (power)", 
       ylab = "Scale Free Topology Model Fit, signed R^2", 
       main = paste("Scale independence"))

text(sft$fitIndices[, 1],
     -sign(sft$fitIndices[, 3]) * sft$fitIndices[, 2],
     labels = powers, cex = cex1, col = "red"
)
abline(h = 0.80, col = "red")
plot(sft$fitIndices[, 1],
     sft$fitIndices[, 5],
     xlab = "Soft Threshold (power)",
     ylab = "Mean Connectivity",
     type = "n",
     main = paste("Mean connectivity")
)
text(sft$fitIndices[, 1],
     sft$fitIndices[, 5],
     labels = powers,
     cex = cex1, col = "red")
```

### Run WGCNA

```{r}
picked_power <- 30 #Max of 30
temp_cor <- cor
cor <- WGCNA::cor
netwk <- blockwiseModules(input_mat,
                           # == Adjacency Function ==
                          power = picked_power,               
                          networkType = "signed",

                          # == Tree and Block Options ==
                          deepSplit = 2,
                          pamRespectsDendro = F,
                          # detectCutHeight = 0.75,
                          minModuleSize = 30,
                          maxBlockSize = 4000,

                          # == Module Adjustments ==
                          reassignThreshold = 0,
                          mergeCutHeight = 0.25,

                          # == TOM == Archive the run results in TOM file (saves time)
                          saveTOMs = T,
                          saveTOMFileBase = "ER",

                          # == Output Options
                          numericLabels = T,
                          verbose = 3,
                          
                          corFnc = "cor",
                          corOptions = list(method = "spearman"))
                          
cor <- temp_cor
```

### Visualize Modules

```{r}
# Convert labels to colors for plotting
mergedColors = labels2colors(netwk$colors)
##############Plot the dendrogram and the module colors underneath
plotDendroAndColors(
  netwk$dendrograms[[1]],
  mergedColors[netwk$blockGenes[[1]]],
  "Module colors",
  dendroLabels = FALSE,
  hang = 0.03,
  addGuide = TRUE,
  guideHang = 0.05 )

module_df <- data.frame(
  gene_id = names(netwk$colors),
  colors = labels2colors(netwk$colors)
)

write_delim(module_df,
            file = "gene_modules.txt",
            delim = "\t")

################Plot Heat Plot
# Get Module Eigengenes per cluster
MEs0 <- moduleEigengenes(input_mat, mergedColors)$eigengenes

# Reorder modules so similar modules are next to each other
MEs0 <- orderMEs(MEs0)
module_order = names(MEs0) %>% gsub("ME","", .)

# Add treatment names
MEs0$treatment = row.names(MEs0)

MEs0$condition <- sub("_[0-9]+$", "", rownames(MEs0))
ME_cols <- grep("^ME", names(MEs0), value = TRUE)

tukey_results <- lapply(ME_cols, function(mod) {
  df <- data.frame(
    eigengene = MEs0[[mod]],
    condition = MEs0$condition
  )
  
  aov_res <- aov(eigengene ~ condition, data = df)
  tukey_res <- TukeyHSD(aov_res)
  
  # Return a tidy data frame with module name attached
  tukey_df <- as.data.frame(tukey_res$condition)
  tukey_df$Comparison <- rownames(tukey_df)
  tukey_df$Module <- mod
  rownames(tukey_df) <- NULL
  tukey_df
})

tukey_results_DF <- do.call(rbind, tukey_results)

SDMadComp <- tukey_results_DF %>%
  filter(grepl("^SDMad-R16", Comparison)) %>%
  transmute(Module, condition = "SDMad", difference = diff, padj = `p adj`)

SD5Comp <- tukey_results_DF %>%
  filter(grepl("^SD5-R16", Comparison)) %>%
  transmute(Module, condition = "SD5", difference = diff, padj = `p adj`)
```

### Generate Module Lists

```{r}
module_df <- data.frame(
  gene_id = names(netwk$colors),
  module = labels2colors(netwk$colors)
)

write_tsv(module_df, "supplementalTables/WGCNAModules.tsv")

# Activated in drive genotypes
A_Genes <- module_df$gene_id[module_df$module == "yellow"]
H_Genes <- module_df$gene_id[module_df$module == "blue"]

# Deactivated in drive genotypes
B_Genes <- module_df$gene_id[module_df$module == "turquoise"]

# Other Modules
C_Genes <- module_df$gene_id[module_df$module == "red"] #Downregulated in SD5
G_Genes <- module_df$gene_id[module_df$module == "brown"] #Downregulated in SDMad
F_Genes <- module_df$gene_id[module_df$module == "green"] #Down SD5, Up SDMad

library(clipr)
write_clip(testis_Background)
write_clip(A_Genes[grepl("^FBgn", A_Genes) == TRUE])
write_clip(H_Genes[grepl("^FBgn", H_Genes) == TRUE])

write_clip(B_Genes[grepl("^FBgn", B_Genes) == TRUE])

write_clip(C_Genes[grepl("^FBgn", C_Genes) == TRUE])
write_clip(F_Genes[grepl("^FBgn", F_Genes) == TRUE])
write_clip(G_Genes[grepl("^FBgn", G_Genes) == TRUE])
```

# Summary Data

```{r}
print("SDMad/Gla vs R16/Gla")
length(na.omit(SDMad_R16_totRNA_Gla$Name[SDMad_R16_totRNA_Gla$p.Adj <= 0.01])) / length(na.omit(SDMad_R16_totRNA_Gla$Name))
length(na.omit(SDMad_R16_totRNA_Repeats_Gla$Name[SDMad_R16_totRNA_Repeats_Gla$p.Adj <= 0.01 & grepl("^FBgn", SDMad_R16_totRNA_Repeats_Gla$Name) == FALSE])) / length(na.omit(SDMad_R16_totRNA_Repeats_Gla$Name[grepl("^FBgn", SDMad_R16_totRNA_Repeats_Gla$Name) == FALSE]))

print("SD5/Gla vs R16/Gla")
length(na.omit(SD5_R16_totRNA_Gla$Name[SD5_R16_totRNA_Gla$p.Adj <= 0.01])) / length(na.omit(SD5_R16_totRNA_Gla$Name))
length(na.omit(SD5_R16_totRNA_Repeats_Gla$Name[SD5_R16_totRNA_Repeats_Gla$p.Adj <= 0.01 & grepl("^FBgn", SD5_R16_totRNA_Repeats_Gla$Name) == FALSE])) / length(na.omit(SD5_R16_totRNA_Repeats_Gla$Name[grepl("^FBgn", SD5_R16_totRNA_Repeats_Gla$Name) == FALSE]))

# Percent differentially expressed annotations that are genes
length(na.omit(SDMad_R16_totRNA_Gla$Name[SDMad_R16_totRNA_Gla$p.Adj < 0.01 & grepl("^FBgn", SDMad_R16_totRNA_Gla$Name) == TRUE ])) / length(na.omit(SDMad_R16_totRNA_Gla$Name[SDMad_R16_totRNA_Gla$p.Adj < 0.01]))

length(na.omit(SD5_R16_totRNA_Gla$Name[SD5_R16_totRNA_Gla$p.Adj < 0.01 & grepl("^FBgn", SD5_R16_totRNA_Gla$Name) == TRUE ])) / length(na.omit(SD5_R16_totRNA_Gla$Name[SD5_R16_totRNA_Gla$p.Adj < 0.01]))
```